share/doc/revng/references/abi-definition.md

```diff
@@ -0,0 +1,213 @@
+- the _second_ GPR register for the `signed` argument.
+- the _third_ vector register for the `float` argument.
+- the _forth_ GPR register for the `unsigned` argument.
+
+If the ABI is NOT position based, the arguments would be passed in:
+- memory for the `Big` argument.
+- the _first_ GPR register for the `signed` argument.
+- the _first_ vector register for the `float` argument.
+- the _second_ GPR register for the `unsigned` argument.
+
+A typical example of a position-based ABI is `Microsoft_x86_64`, a non-position-based one - `SystemV_x86_64`.
+
+## `OnlyStartDoubleArgumentsFromAnEvenRegister` (`bool`)
+
+States whether an object that needs two GPRs to fit (5-8 bytes on 32-bit architectures and 9-16 bytes on 64-bit systems) is only allowed to start from a register with an even index.
+
+Here's an example function:
+```
+void function(uint32_t, uint64_t);
+```
+On a system with 32-bit wide GPRs, the first argument (`uint32_t`) is passed using the first allowed GPR (say `r0`).
+
+The second argument (`uint64_t`) requires two register to fit, so it's passed using second and third registers if the ABI allows starting double arguments from any register (`r1` and `r2` in this example), or third and forth if it only allows starting them from even registers (`r2` and `r3` in this example, since `r1` is considered an odd register (the count starts from 0, much like C array indexing).
+
+## `ArgumentsCanBeSplitBetweenRegistersAndStack` (`bool`)
+
+States whether the ABI allows a single object that wouldn't fit into a single GPR (9+ bytes on 32-bit systems and 17+ bytes on 64-bit ones) to be partially passed in registers with the remainder placed on the stack if there are not enough registers to fit the entirety of it.
+
+As an example, let's say that there is a big object of type `Big` such that `sizeof(Big)` is equal to 16 bytes. On 32-bit system it would mean having to use four GPRs (`16 == 4 * 4`) to fit it.
+
+Let's look at an ABI that allocates four registers for passing function arguments (`r0-r3`). Then, for a function like
+```cpp
+void function(uint32_t, Big);
+```
+the `uint32_t` argument would be passed in the first GPR (`r0`).
+
+But that would also mean that the remaining three available GPRs are not enough to fit the entirety of the `Big` object, meaning it needs to either be split between the registers and the memory, or passed using the stack. That's exactly what this option states.
+
+## `BigArgumentsUsePointersToCopy` (`bool`)
+
+This allows a non-position-based ABI to take advantage of one of the main traits of position-based ABI: ability to use pointers to copy to pass big aggregates.
+
+This is one of the defining features of arm64 ABI, which allows it to get the best of both approaches - an ability to pass aggregates by copy in registers from the non-position-based approach while also being able to sometimes avoid a copy of a bigger aggregate from the position-based approach.
```

Comment:
> also being able to sometimes avoid a copy of a bigger aggregate from the position-based approach

You seem to imply this saves some copying, but in practice this will still be a pointer to a copy, I think, since the callee can change it.

Reply:
Well, that's usually the point. If it can prove that the callee never touches it, that copy is the very first thing to go away.

Reply:
By "never touches it" I guess you mean doesn't write to it.

You're right, but I'm not sure that was the original design goal, it's a rather advanced compiler optimization that requires view on the callee, seems weird for an ABI to be designed for such a thing.

It'd be interesting to find a counterexample on godbolt, but it's not too important.
I'll leave this to you.

Reply:
> By "never touches it" I guess you mean doesn't write to it.

Yeah, sorry.

------------

I looked around but couldn't find source I originally referred too. So I went and played around in godbolt.

I couldn't make it happen under msvc at all. It kept trying to inline the callee. And explicitly forbidding it from doing so seems to disable optimizations around the call in the first place (it's probably similar to gcc's `weak` attribute) - so not much success there.

I did manage to reproduce the behavior with [clang](https://godbolt.org/z/EETvKEKGK) using only `-O1`, but again, who knows how much of that is just clang being awesome.

That being said, I think you're right and we should just drop this from docs.

Reply:
```
This is one of the defining features of arm64 ABI, which allows it to get the best of both approaches - an ability to pass aggregates by copy in registers from the non-position-based approach while also being able to avoid "wasting" registers on passing big aggregates from the position-based approach.
```

?

Reply:
Works.

Reply:
Done.